archive/imo1988_q6.lean

```diff
@@ -0,0 +1,271 @@
+  -- Note that the point p = (x,y) lies on the upper branch.
+  let p : ℕ × ℕ := ⟨x,y⟩,
+  have hp : p ∈ upper_branch := ⟨h₀, hxy⟩,
+  -- We also consider the exceptional set of solutions (a,b) that satisfy
+  -- a = 0 or a = b or B a = b or B a = b + a.
+  let exceptional  : set (ℕ × ℕ) :=
+  {p | H p.1 p.2 ∧ (base p.1 p.2 ∨ p.1 = 0 ∨ p.1 = p.2 ∨ B p.1 = p.2 ∨ B p.1 = p.2 + p.1) },
+  -- Let S be the projection on to the y-axis of the upper branch after removing the exceptional locus.
+  let S : set ℕ := prod.snd '' (upper_branch \ exceptional),
+  -- The strategy is to show that the exceptional locus in nonempty.
+  -- We first show that this is sufficient.
+  suffices exc : exceptional ≠ ∅,
+  { -- Suppose that there exists an element in the exceptional locus.
+    rw set.ne_empty_iff_exists_mem at exc,
+    simp [exceptional, -add_comm] at exc,
+    -- Let (a,b) be such an element, and consider all the possible cases.
+    rcases exc with ⟨a, b, hH, hb⟩, rcases hb with _|rfl|rfl|hB|hB,
+    -- The first three cases are rather easy to solve.
+    { solve_by_elim },
+    { rw H_symm at hH, solve_by_elim },
+    { solve_by_elim },
+    -- The final two cases are very similar.
+    all_goals {
+      -- Consider the quadratic equation that (a,b) satisfies.
+      rw H_quad at hH,
+      -- We find the other root of the equation, and Vieta's formulas.
+      rcases Vieta_formula_quadratic hH with ⟨c, h_root, hV₁, hV₂⟩,
+      -- By substitutions we find that b = 0 or b = a.
+      simp [hB] at hV₁, subst hV₁,
+      rw [← int.coe_nat_zero] at *,
+      rw ← H_quad at h_root,
+      -- And hence we are done by H_zero and H_diag.
+      solve_by_elim } },
+  -- Now assume that the exceptional locus is empty.
+  assume exceptional_empty,
+  -- Observe that S is nonempty.
+  have S_ne_empty : S ≠ ∅,
+  { rw set.ne_empty_iff_exists_mem,
+    -- It contains the image of p.
+    use p.2,
+    apply set.mem_image_of_mem,
+    -- After all, we assumed that the exceptional locus is empty.
+    rwa [exceptional_empty, set.diff_empty], },
+  -- We are now set for an infinite descent argument.
+  -- Let m be the smallest element of the nonempty set S.
+  let  m     : ℕ                := well_founded.min     nat.lt_wf S S_ne_empty,
+  have m_mem : m ∈ S            := well_founded.min_mem nat.lt_wf S S_ne_empty,
+  have m_min : ∀ k ∈ S, ¬ k < m := λ k hk, well_founded.not_lt_min nat.lt_wf S S_ne_empty hk,
+  -- It suffices to show that there is point (a,b) with b ∈ S and b < m.
+  suffices hp' : ∃ p' : ℕ × ℕ, p'.2 ∈ S ∧ p'.2 < m,
+  { rcases hp' with ⟨p', p'_mem, p'_small⟩, solve_by_elim },
+  -- Let (m_x, m_y) be a point on the upper branch that projects to m ∈ S
+  -- and that does not lie in the exceptional locus.
+  rcases m_mem with ⟨⟨mx, my⟩, ⟨⟨hHm, mx_lt_my⟩, h_base⟩, m_eq⟩,
+  -- This means that m_y = m,
+  -- and the conditions H(m_x, m_y) and m_x < m_y are satisfied.
+  simp only at hHm,
```

Comment:
What does `simp only` do if you leave out the list?

Reply:
It does some very minor definitional reductions and such. I had it here for presentational purposes. But I can leave it out.